Identification:

```python
name = ["YURI REINALDO DA SILVA", "FERNANDO RODRIGUES MACIEL"]
id   = [2015037403, 20150127906]
```



<img width="60" src="https://drive.google.com/uc?export=view&id=1JQRWCUpJNAvselJbC_K5xa5mcKl1gBQe"> 



In [2]:
# Uploading files from your local file system

from google.colab import files
uploaded = files.upload()
for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving autos.csv to autos.csv
User uploaded file "autos.csv" with length 9617699 bytes


In [3]:
!ls

autos.csv  sample_data


# 1.0 Introduction

In this lesson, we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). We've made a few modifications from the original dataset that was uploaded to Kaggle:

- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:


- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which which year the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which which year the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- Start by writing a paragraph in a markdown cell introducing the project and the dataset.
- Import the pandas and NumPy libraries
- Read the **autos.csv** CSV file into pandas, and assign it to the variable name **autos**.
    - Try without specifying any encoding (which will default to **UTF-8**)
    - If you get an encoding error, try the next two most popular encodings (**Latin-1** and **Windows-1252**) until you are able to read the file without error.
- Create a new cell with just the variable **autos** and run this cell.
A neat feature of jupyter notebook is its ability to render the first few and last few values of any pandas object.
- Use the **DataFrame.info()** and **DataFrame.head()** methods to print information about the **autos** dataframe, as well as the first few rows.
- Write a markdown cell briefly describing your observations.



## Apresentação do projeto

Durante este projeto estaremos analisando os dados obtidos pela seção de classificados do site *eBay Kleinanzeigen*, site alemão com foco em anúncios para venda de veículos usados. 

No decorrer deste notebook estaremos realizando o tratamento dos dados e análises sobre os resultados encontrados.

In [0]:
# put your code here
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [7]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [8]:
print(autos['fuelType'].unique())

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']


### Write a markdown cell briefly describing your observations.

- Padrão UTF-8 causa falhas em caracteres especiais
- A coluna nrOfPictures apresenta todos os valores como 0
- Quase todas as colunas tem valores nulos, se apresentando como NaN ou 0



# 2.0 Cleaning Column Names



From the work we did in the last section, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have **null** values, but none have more than ~20% **null** values.
- The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

Let's convert the column names from **camelcase** to **snakecase** and reword some of the column names based on the data dictionary to be more descriptive.

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">

- Use the **DataFrame.columns** attribute to print an array of the existing column names.
- Copy that array and make the following edits to columns names:
    - **yearOfRegistration** to **registration_year**
    - **monthOfRegistration** to **registration_month**
    - **notRepairedDamage** to **unrepaired_damage**
    - **dateCreated** to **ad_created**
    - The rest of the columnn names from camelcase to snakecase.
- Assign the modified column names back to the **DataFrame.columns** attribute.
- Use **DataFrame.head()** to look at the current state of the autos dataframe.
- Write a markdown cell explaining the changes you made and why.

In [9]:
# put your code here
print(autos.columns)

import re

_underscorer1 = re.compile(r'(.)([A-Z][a-z]+)')
_underscorer2 = re.compile('([a-z0-9])([A-Z])')

def camelToSnake(s):
  subbed = _underscorer1.sub(r'\1_\2', s)
  return _underscorer2.sub(r'\1_\2', subbed).lower()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [10]:
col_names = {
    "yearOfRegistration" : "registration_year",
    "monthOfRegistration" : "registration_month",
    "notRepairedDamage" : "unrepaired_damage",
    "dateCreated" : "ad_created"
}

for col in autos.columns:
  if col not in col_names:
    col_names[col] = camelToSnake(col)

    
print(col_names)

{'yearOfRegistration': 'registration_year', 'monthOfRegistration': 'registration_month', 'notRepairedDamage': 'unrepaired_damage', 'dateCreated': 'ad_created', 'dateCrawled': 'date_crawled', 'name': 'name', 'seller': 'seller', 'offerType': 'offer_type', 'price': 'price', 'abtest': 'abtest', 'vehicleType': 'vehicle_type', 'gearbox': 'gearbox', 'powerPS': 'power_ps', 'model': 'model', 'odometer': 'odometer', 'fuelType': 'fuel_type', 'brand': 'brand', 'nrOfPictures': 'nr_of_pictures', 'postalCode': 'postal_code', 'lastSeen': 'last_seen'}


In [11]:
autos.rename(columns = col_names, inplace=True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Alterações nos nomes das colunas

- Fizemos alterações nos nomes das colunas para que ficasse mais legível e de forma que todos sigam o modelo snakecase, que é o mais adotado para Python

# 3.0 Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data:
- **DataFrame.describe()** (with **include='all'** to get both categorical and numeric columns) 
- **Series.value_counts()** and **Series.head()** if any columns need a closer look.

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">

- Use **DataFrame.describe()** to look at descriptive statistics for all columns.
- Write a markdown cell noting:
  - Any columns that have mostly one value that are candidates to be dropped
  - Any columns that need more investigation.
  - Any examples of numeric data stored as text that needs to be cleaned.
- If you need to investigate any columns more, do so and write up any additional things you found.
- You likely found that the **price** and **odometer** columns are numeric values stored as text. For each column:
    - Remove any non-numeric characters.
    - Convert the column to a numeric dtype.
    - Use **DataFrame.rename()** to rename the column to **odometer_km**.

In [12]:
# put your code here
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

## Há algumas coisas erradas aqui
Analisando as informações dadas pelo método *Dataframe.describe()* dá para se reparar algumas coisas inconsistentes com a realidade:

- Os valores mínimos e máximos estão além da faixa comum para algumas colunas.
  - No caso de *registration_year* temos que o mínimo é 1000 e o máximo é 9999, que são inválidos.
  - A coluna *power_ps* tem valores 0 e 17.700, que são impossíveis pois todo carro tem uma potência e o carro comercial de maior potência possui 808 CVs.
  - Em *registration_month* temos que o menor valor é 0, que também é inválido.
- Os valores das colunas *Price* e *Odometer* podem ser salvos como números removendo os caracteres redundantes.
- A coluna *nrOfPictures* só tem valores 0 e pode ser removida



In [13]:
# Transformando a coluna price em inteiros

autos['price'] = [x.replace('$', '').replace(',', '')
                  for x in autos['price']]
pd.to_numeric(autos['price'], errors='ignore')
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000  control          bus               2004   
1  privat    Angebot  8500  control    limousine               1997   
2  privat    Angebot  8990     test    limousine               2009   
3  privat    Angebot  4350  control   kleinwagen               2007   
4  privat    Angebot  1350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [0]:
# Transformando a coluna price em inteiros
autos["odometer"] = [x.replace(",", "").replace("km", "") for x in autos["odometer"]]
autos.rename(columns= {"odometer" : "odometer_km"}, inplace = True)

In [0]:
autos["price"] = autos["price"].astype(int)
autos["odometer_km"] = autos["odometer_km"].astype(int)

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int64
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(7), 

# 4.0 Exploring the Odometer and Price Columns

From the last section, we learned that there are a number of text columns where almost all of the values are the same (**seller** and **offer_type**). We also converted the **price** and **odometer** columns to numeric types and renamed **odometer** to **odometer_km**.

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the **odometer_km** and **price** columns. Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
- We'll use:
  - **Series.unique().shape** to see how many unique values
  - **Series.describe()** to view min/max/median/mean etc
  - **Series.value_counts()**, with some variations:
    - chained to **.head()** if there are lots of values.
    - Because **Series.value_counts()** returns a series, we can use **Series.sort_index()** with **ascending=True** or **False** to view the highest and lowest values with their counts (can also chain to head() here).
    - When removing outliers, we can do df[(df["col"] > x ) & (df["col"] < y )], but it's more readable to use df[df["col"].[between(x,y)](https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.Series.between.html)]
    

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- For each of the **odometer_km** and **price** columns:
  - Use the techniques above to explore the data
  - If you find there are outliers, remove them and write a markdown paragraph explaining your decision.
  - After you have removed the outliers, make some observations about the remaining values, with examples of how we can use that to extract **YYYY**, **YYYY-MM** or **YYYY-MM-DD** from those columns.

In [17]:
# put your code here
print("Status for odometer_km column\n")
print("Unique values : ", autos["odometer_km"].unique().shape[0], "\n")
print("Series description : ", autos["odometer_km"].describe(), "\n")
print("Value occurrences : ", autos["odometer_km"].value_counts(ascending=False), "\n")

Status for odometer_km column

Unique values :  13 

Series description :  count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

Value occurrences :  150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64 



In [18]:
print("Unique values : ", autos["price"].unique().shape[0], "\n")
print("Value occurrences : ", autos["price"].value_counts(ascending=False), "\n")
print("Series description : ", autos["price"].describe(), "\n")

Unique values :  2357 

Value occurrences :  0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64 

In [0]:
# Removing outliers
odometer_min = 125000.000000
odometer_max = 150000.000000

odometer_iq = 1.5 *  (odometer_max - odometer_min)

price_min = 1.100000e+03
price_max = 7.200000e+03

price_iq = 1.5 * (price_max - price_min)
 

In [23]:
autos = autos[autos["odometer_km"].between(odometer_min - odometer_iq, odometer_max + odometer_iq)]

autos = autos[autos["price"].between(price_min - price_iq, price_max + price_iq)]

autos.shape

(39964, 20)

In [25]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  39964.000000       39964.000000  39964.000000   39964.000000   
mean    3558.803899        2002.675083    109.948253  142286.557902   
std     3536.038723          35.644961    194.010425   16537.626480   
min        0.000000        1910.000000      0.000000   90000.000000   
25%      990.000000        1999.000000     68.000000  150000.000000   
50%     2299.000000        2002.000000    105.000000  150000.000000   
75%     4999.000000        2006.000000    141.000000  150000.000000   
max    16350.000000        9000.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        39964.000000         39964.0  39964.000000  
mean             5.698754             0.0  50408.937168  
std              3.740147             0.0  25620.312782  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  29699.000000  
50%              6.000000             0.0  49090.000000  
75%              9.000000             0.0  70771.000000  
max             12.000000             0.0  99998.000000

## Após remoção dos valores questionáveis
Com a remoção das linhas com outliers nas colunas *price* e *odometer_km* o dataset teve uma diminuição de cerca de 10.000 elementos. O que pode ser considerado uma mudança drástica, mas que ajdua a obtermos informações mais precisas já que removemos dados impuros. 

Após essas mudanças podemos ver que os valores de média e desvio padrão ficaram mais normalizados, pois extavam sendo alterados pelos valores extremos

# 5.0 Exploring the date columns



Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:


- **date_crawled**: added by the crawler
- **last_seen**: added by the crawler
- **ad_created**: from the website
- **registration_month**: from the website
- **registration_year**: from the website

Right now, the **date_crawled**, **last_seen**, and **ad_created** columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like **Series.describe()** to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

```python
autos[['date_crawled','ad_created','last_seen']][0:5]
```

|  |date_crawled | ad_created          | last_seen           |                     |
|--------------|---------------------|---------------------|---------------------|
| 0            | 2016-03-26 17:47:46 | 2016-03-26 00:00:00 | 2016-04-06 06:45:54 |
| 1            | 2016-04-04 13:38:56 | 2016-04-04 00:00:00 | 2016-04-06 14:45:08 |
| 2            | 2016-03-26 18:57:24 | 2016-03-26 00:00:00 | 2016-04-06 20:15:37 |
| 3            | 2016-03-12 16:58:10 | 2016-03-12 00:00:00 | 2016-03-15 03:16:28 |
| 4            | 2016-04-01 14:38:50 | 2016-04-01 00:00:00 | 2016-04-01 14:38:50 |


You'll notice that the first 10 characters represent the day (e.g. **2016-03-12**). To understand the date range, we can extract just the date values, use **Series.value_counts()** to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use **Series.str[:10]**:

```python
autos['date_crawled'].str[:10]

>>>
0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
...
```

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- Use the workflow we just described to calculate the distribution of values in the **date_crawled**, **ad_created**, and **last_seen** columns (all string columns) as percentages.
  - To include missing values in the distribution and to use percentages instead of counts, chain the **Series.value_counts(normalize=True, dropna=False)** method.
  - To rank by date in ascending order (earliest to latest), chain the **Series.sort_index()** method.
  - Write a markdown cell after each column exploration to explain your observations.
- Use **Series.describe()** to understand the distribution of **registration_year**.
    - Write a markdown cell explaining your observations.

In [30]:
# put your code here
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.024947
2016-03-06    0.014063
2016-03-07    0.036283
2016-03-08    0.034231
2016-03-09    0.033805
2016-03-10    0.032579
2016-03-11    0.032079
2016-03-12    0.037384
2016-03-13    0.015014
2016-03-14    0.037158
2016-03-15    0.034006
2016-03-16    0.029852
2016-03-17    0.031954
2016-03-18    0.013187
2016-03-19    0.033580
2016-03-20    0.037309
2016-03-21    0.037534
2016-03-22    0.032504
2016-03-23    0.032654
2016-03-24    0.029001
2016-03-25    0.032629
2016-03-26    0.032454
2016-03-27    0.030928
2016-03-28    0.034631
2016-03-29    0.033830
2016-03-30    0.033981
2016-03-31    0.031779
2016-04-01    0.033055
2016-04-02    0.034706
2016-04-03    0.038184
2016-04-04    0.036733
2016-04-05    0.013362
2016-04-06    0.003453
2016-04-07    0.001151
Name: date_crawled, dtype: float64

## Data de rastreamento dos anuncios

Nesta coluna podemos ver que o rastreamento dos anuncios foi feito de forma semi-uniforme no período entre 2016-03-05 e 2016-04-05 com cada dia tendo uma quantidade similar de rastreios. A exceção é que nos ultimos 2 dias escritos (2016-04-06   e
2016-04-07) foram feitos bem menos rastreios em comparação com os anteriores.

In [31]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-12-05    0.000025
2015-12-30    0.000025
2016-01-03    0.000025
2016-01-07    0.000025
2016-01-10    0.000050
2016-01-13    0.000025
2016-01-16    0.000025
2016-01-22    0.000025
2016-01-27    0.000050
2016-01-29    0.000025
2016-02-01    0.000025
2016-02-02    0.000050
2016-02-05    0.000050
2016-02-08    0.000025
2016-02-09    0.000025
2016-02-11    0.000025
2016-02-12    0.000075
2016-02-16    0.000025
2016-02-18    0.000050
2016-02-19    0.000025
2016-02-20    0.000050
2016-02-21    0.000075
2016-02-22    0.000025
2016-02-23    0.000025
2016-02-24    0.000025
2016-02-25    0.000050
2016-02-26    0.000025
2016-02-27    0.000125
2016-02-28    0.000200
2016-02-29    0.000150
                ...   
2016-03-09    0.033680
2016-03-10    0.032404
2016-03-11    0.032379
2016-03-12    0.037234
2016-03-13    0.016465
2016-03-14    0.035607
2016-03-15    0.033805
2016-03-16    0.030302
2016-03-17    0.031653
2016-03-18    0.013887
2016-03-19    0.032579
2016-03-20    0.037309
2016-03-21 

## Data de criação dos anúncios

Já na coluna *ad_created* temos datas entre 2015-12-05 e 2016-04-01. Sendo que até 2016-03 há poucos anuncios sendo criados, portanto a maioria dos dados provem desse mês em diante.

In [34]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001126
2016-03-06    0.004804
2016-03-07    0.006005
2016-03-08    0.008558
2016-03-09    0.010560
2016-03-10    0.011210
2016-03-11    0.013787
2016-03-12    0.025598
2016-03-13    0.009609
2016-03-14    0.013412
2016-03-15    0.016590
2016-03-16    0.017441
2016-03-17    0.029777
2016-03-18    0.007507
2016-03-19    0.016890
2016-03-20    0.022020
2016-03-21    0.021770
2016-03-22    0.022370
2016-03-23    0.019442
2016-03-24    0.020844
2016-03-25    0.020218
2016-03-26    0.017666
2016-03-27    0.016890
2016-03-28    0.022045
2016-03-29    0.023046
2016-03-30    0.025448
2016-03-31    0.024597
2016-04-01    0.024222
2016-04-02    0.025673
2016-04-03    0.025623
2016-04-04    0.025323
2016-04-05    0.118457
2016-04-06    0.209063
2016-04-07    0.122410
Name: last_seen, dtype: float64

## Anuncios vistos pela ultima vez

Analisando os dados, podemos afirmar que uma parcela considerável dos anúncios continua sendo visualizado com o passar do tempo. Cerca de 43% dos anuncios foram visualizados nesses ultimos 3 dias resgistrados.

In [35]:
autos["registration_year"].describe()

count    39964.000000
mean      2002.675083
std         35.644961
min       1910.000000
25%       1999.000000
50%       2002.000000
75%       2006.000000
max       9000.000000
Name: registration_year, dtype: float64

## Ano de registro do veículo

O ano de registro possui dados inválidos que corrompem a sua análise, o valor máximo encontrado é 9000, ou seja muitos anos a frente no futuro. Como há valores extremos os calculos de média e desvio padrão ficam falhos e não condizentes com a realidade.

# 6.0 Dealing with Incorrect Registration Year Data


One thing that stands out from the exploration we did in the last section is that the **registration_year** column contains some odd values:

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.


**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- Decide which the highest and lowest acceptable values are for the **registration_year** column.
  - Write a markdown cell explaining your decision and why.
- Remove the values outside those upper and lower bounds and calculate the distribution of the remaining values using **Series.value_counts(normalize=True)**.
  - Write a markdown cell explaining your observations.


In [0]:
# put your code here
reg_year_min = 1920
reg_year_max = 2016

## Ano de registro mínimo e máximo
Optamos por 1920 para o ano mínimo porque é o ano seguinte da grande depressão em que muitas marcas que existiam fecharam as portas e também nos dá uma faixa de cerca de 100 anos para os veículos anunciados.

O ano máximo é obviamente o de 2016 pois é o ano em que os anuncios foram produzidos, não se pode vender carros muito futuros como usados.

In [45]:
autos = autos[autos["registration_year"].between(reg_year_min, reg_year_max)]

autos["registration_year"].value_counts(normalize=True).sort_index()

1934    0.000026
1937    0.000026
1950    0.000026
1953    0.000026
1954    0.000052
1956    0.000052
1958    0.000052
1959    0.000026
1960    0.000392
1961    0.000052
1962    0.000026
1963    0.000078
1964    0.000052
1965    0.000183
1966    0.000183
1967    0.000183
1968    0.000235
1969    0.000157
1970    0.000418
1971    0.000183
1972    0.000444
1973    0.000418
1974    0.000235
1975    0.000235
1976    0.000392
1977    0.000314
1978    0.000601
1979    0.000549
1980    0.001594
1981    0.000497
          ...   
1987    0.001620
1988    0.003084
1989    0.004051
1990    0.008442
1991    0.008285
1992    0.009017
1993    0.010716
1994    0.015786
1995    0.031180
1996    0.035466
1997    0.050729
1998    0.060399
1999    0.074539
2000    0.081334
2001    0.066593
2002    0.062203
2003    0.066750
2004    0.066201
2005    0.070854
2006    0.061732
2007    0.049631
2008    0.043019
2009    0.032983
2010    0.019811
2011    0.013930
2012    0.006194
2013    0.001228
2014    0.0003

## Anos de registro dos veículos anunciados

Na faixa definida temos que os veículos anunciados vão de 1934 à 2016, sendo muito mais raros os veículos anteriores aos anos 80. A maioria dos anuncios se concentra pelos veículos dos anos 90 e 2000. 

Há uma queda leve nos anos recentes, como a partir de 2012. Por se tratarem de veículos relativamente novos os donos ainda não tem muito interesse de vendê-los como usados. Mas há um número considerável de veículos de 2016 o mesmo ano em que foi feita a análise, isso pode ser causado pelos anunciantes preencherem a mesma data em que o anuncio foi feito por engano.

# 7.0 Exploring Price by Brand



One of the analysis techniques we learned in this course is aggregation. When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the **brand** column.

If you recall in an earlier lesson, we explored how to use loops to perform aggregation. Here's what the process looks like:


- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.
    
    
**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- Explore the unique values in the **brand** column, and decide on which brands you want to aggregate by.
  - You might want to select the top 20, or you might want to select those that have over a certain percentage of the total values (e.g. > 5%).
  - Remember that **Series.value_counts()** produces a series with index labels, so you can use **Series.index** attribute to access the labels, should you wish.
- Write a short paragraph describing the brand data, and explaining which brands you've chosen to aggregate on.
- Create an empty dictionary to hold your aggregate data.
  - Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.
  - Print your dictionary of aggregate data, and write a paragraph analyzing the results.

In [62]:
# put your code here
print(autos["brand"].unique())

brands = autos["brand"].value_counts(normalize=True) 
brand_higher_2percent = brands > 0.02


selected_brands = brands[brand_higher_2percent]

print("\nBrands with more than 2% of the ads announced:")
selected_brands

['peugeot' 'bmw' 'ford' 'chrysler' 'volkswagen' 'seat' 'renault'
 'mercedes_benz' 'audi' 'opel' 'mazda' 'mini' 'toyota' 'nissan' 'jeep'
 'dacia' 'saab' 'volvo' 'mitsubishi' 'fiat' 'skoda' 'subaru' 'kia'
 'sonstige_autos' 'citroen' 'smart' 'porsche' 'hyundai' 'chevrolet'
 'honda' 'daewoo' 'suzuki' 'land_rover' 'jaguar' 'alfa_romeo' 'lada'
 'rover' 'trabant' 'daihatsu' 'lancia']

Brands with more than 2% of the ads announced:


volkswagen       0.221029
opel             0.117401
bmw              0.112514
mercedes_benz    0.093905
audi             0.083948
ford             0.071037
renault          0.051304
peugeot          0.031310
fiat             0.024855
Name: brand, dtype: float64

## Marcas com maior numero de veículos anunciados
Escolhemos as empresas com parcela maior que 2% dos anuncios exibidos, para podermos visualizar bem quais são as principais marcas do mercado.

É possível reparar que o mercado é dominado pelas empresas alemãs, com as 5 primeiras sendo nacionais ao site de anuncios e representando cerca de 60% do total. A empresa Volkswagen possui uma faixa de 22% dos anuncios dessas empresas, o que pode ter sido alcançado por ser uma empresa nacional de carros mais populares.

In [0]:
mean_price_brand = {}

for br in selected_brands.index:
  brand_filter = autos["brand"] == br
  mean_price_brand[br] = autos[brand_filter]["price"].mean()

In [60]:
mean_price_brand

{'audi': 5220.4346201743465,
 'bmw': 5326.838095238095,
 'fiat': 1873.4079915878024,
 'ford': 2389.4621044885944,
 'mercedes_benz': 4969.493181185639,
 'opel': 2180.565449688335,
 'peugeot': 2358.8213689482473,
 'renault': 1759.2878247580234,
 'volkswagen': 3587.68038311458}

## Preço médio das marcas mais populares

Podemos ver que o preço médio varia entre 1.750 e 5.326 euros. Com a BMW se mostrando a marca mais cara e ainda assim permanecendo como a 3ª mais popular do site, isso indica que existe um nicho para veículos mais caros no site. Os carros da Renault são apesar de costumarem ser os mais baratos compõem apenas 5% dos anuncios do site. A Volkswagen sendo a mais popular apresenta uma média de preços razoável com seus veículos custando em média 3.87 euros.



# 8.0 Storing Aggregate Data in a DataFrame

In the last section, we aggregated across brands to understand mean price. We observed that in the top 5 brands, there's a distinct price gap.

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For the top 5 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons


Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

- [pandas series constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html)
- [pandas dataframe constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html)

Here's an example of the series constructor that uses the **brand_mean_prices** dictionary:


```python
bmp_series = pd.Series(brand_mean_prices)
bmp_series

>>>
audi             9336
bmw              8332
ford             3749
mercedes_benz    8628
opel             2975
volkswagen       5402
dtype: int64
```


The keys in the dictionary bceame the index in the series object. We can then create a single-column dataframe from this series object. We need to use the **columns** parameter when calling the dataframe constructor (which accepts a array-like object) to specify the column name (or the column name will be set to 0 by default):

```python
df = pd.DataFrame(bmp_series, columns=['mean_mileage'])
df
```


| | mean_mileage  |      
|---------------|------|
| bmw           | 8332 |
| mercedes_benz | 8628 |
| opel          | 2975 |
| audi          | 9336 |
| volkswagen    | 5402 |
| ford          | 3749 |


**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.
- Convert both dictionaries to series objects, using the series constructor.
- Create a dataframe from the first series object using the dataframe constructor.
- Assign the other series as a new column in this dataframe.
Pretty print the dataframe, and write a paragraph analyzing the aggregate data.




In [69]:
# put your code here
top_6_mean_price = {}
top_6_mean_mile = {}

for br in selected_brands.index[:6]:
  br_filter = autos["brand"] == br
  top_6_mean_price[br] = autos[br_filter]["price"].mean()
  top_6_mean_mile[br] = autos[br_filter]["odometer_km"].mean()
  
print("Mean Prices:")
print(top_6_mean_price)
print("Mean Mileage:")
print(top_6_mean_mile)

Mean Prices:
{'volkswagen': 3587.68038311458, 'opel': 2180.565449688335, 'bmw': 5326.838095238095, 'mercedes_benz': 4969.493181185639, 'audi': 5220.4346201743465, 'ford': 2389.4621044885944}
Mean Mileage:
{'volkswagen': 143644.90954239093, 'opel': 142087.04363312555, 'bmw': 144838.55981416957, 'mercedes_benz': 145001.3915947676, 'audi': 145926.21419676213, 'ford': 140794.70198675495}


In [70]:
top_6_price_series = pd.Series(top_6_mean_price)
top_6_mile_series = pd.Series(top_6_mean_mile)

print(top_6_price_series)
print(top_6_mile_series)

audi             5220.434620
bmw              5326.838095
ford             2389.462104
mercedes_benz    4969.493181
opel             2180.565450
volkswagen       3587.680383
dtype: float64
audi             145926.214197
bmw              144838.559814
ford             140794.701987
mercedes_benz    145001.391595
opel             142087.043633
volkswagen       143644.909542
dtype: float64


In [71]:
top_6_df = pd.DataFrame(top_6_price_series, columns={"price"})

top_6_df["mileage_km"] = top_6_mile_series

top_6_df

price     mileage_km
audi           5220.434620  145926.214197
bmw            5326.838095  144838.559814
ford           2389.462104  140794.701987
mercedes_benz  4969.493181  145001.391595
opel           2180.565450  142087.043633
volkswagen     3587.680383  143644.909542

## Analisando o dataframe

Com o dataframe podemos ver que apesar de serem mais caros, os veículos das marcas mais caras (Audi, BMW e Mercedes Benz) tem uma kilometragem maior antes de serem vendidos. Enquanto as marcas mais baratas (Ford e Opel) apresentam as menores kilometragens médias.

Há uma forte correlação entre o preço e a kilometragem. Isso pode ser indicativo que quanto mais caro o veículo, menor a rotatividade que seus donos tem de vendê-los após uso. 